### ISSUE TITLE
Kubernetes - Cloud Specific Integrations (aws, gce, openstack)

### ISSUE THREAD

**Issue by eddytruyen** [mbruzek/layer-k8s#42]
If I deploy the kubernetes bundle in an openstack juju environment, will the kubernetes api-server and controller be configured with the options `--cloud-provider=openstack` and `--cloud-config=<file on hostpath of hyperkube image containing openstack configuration parameters from environments.yaml such as tenant-id>`

As such I it is possible to use the cinder volume plugin of kubernetes.

Best regards,

Eddy

**Comment by eddytruyen** [comment]
I am closing this question because I already figured out the answer. It is not really possible because that would need to expose sensitive information such as my passwd at openstack.

**Comment by lazypower** [comment]
@eddytruyen  We do want to enable the cloud specific componentry of kubernetes however.

I've been thinking about this, and have discerned we can do it in a couple of ways:
- Charm Config (gets a bit convoluted)
- Provide the configuration via juju-resources as a yaml file

And i'm sure there are other options I haven't called out. However having the capacity to deploy kubernetes and get load balancers deployed on GCE/AWS/OpenStack would be great.

**Comment by eddytruyen** [comment]
Yes I agree. My last closing comment about password-leakage was not correct. Somehow I was thinking in the business model of juju-as-a-service when writing this.

In the docker-multi-node project I experimented with adding these options in the master.sh file. 
This worked for `--cloud-config=openstack` but not for setting  the `cloud-config` option. Apparently the value of this option must be a file that is part of the hyperkube image or mounted into the kubelet container. I don't know how to do that in a generic way.

**Comment by lazypower** [comment]
Right, this will need to be passed to hyperkube. I think we should re-open this issue and re-target it.

**Comment by lazypower** [comment]
@mbruzek  -  We should probably try to fit this in our next maintenance window later this cycle. Its distributing keys and then validations from what i can tell. Looks fairly straight forward (Famous last words)

**Comment by eddytruyen** [comment]
More information about this can be found here: http://kubernetes.io/docs/getting-started-guides/scratch/#cloud-providers

**Comment by eddytruyen** [comment]
It looks that the openstack integration in Kubernetes is buggy.

This weekend I tried to get the openstack integration in the kubernetes project to work in my openstack environment. I have tried to do this in several kubernetes deployments. Hyperkubev1.1.7, Hyperkubev1.2.4, and the recent openstack-heat provider in Kubernetes release v.1.2.4. None of them worked. I could not attach cinder volumes and the LoadBalancer servicetype worked neither.

This is just to say that it is not worth the effort of trying to fit this into juju, at least for now.
